### PR TITLE
GEODE-6927 make getThreadOwnedConnection code thread safe

### PR DESCRIPTION
**Pull request by mkevo**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java [hunk 1]
```diff
@@ -494,20 +494,28 @@ Connection getThreadOwnedConnection(DistributedMember id, long startTime, long a
...
+      if (this.threadConnectionMap == null) {
+        // This instance is being destroyed; fail the operation
+        closeCon(
+            "Connection table being destroyed",
+            result);
+        return null;
+      }
       // Since it's a concurrent map, we just try to put it and then
       // return whichever we got.
       Object o = this.threadConnectionMap.putIfAbsent(id, al);
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java]: A simplified version of this looks more like:
```java
...
final Map local = this.member;
if (null == local) {
  return null;
}
final List list = local.computeIfAbsent(id, ArrayList::new);
...
```


Use `JavaWorkarounds.computeIfAbsent` to avoid the penalty of `computeIfAbsent`.

File: geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java [hunk 2]
```diff
@@ -494,20 +494,28 @@ Connection getThreadOwnedConnection(DistributedMember id, long startTime, long a
...
     }
 
-    ArrayList al = (ArrayList) this.threadConnectionMap.get(id);
+    ArrayList al = (ArrayList) cm.get(id);
     if (al == null) {
       // First connection for this DistributedMember. Make sure list for this
       // stub is created if it isn't already there.
       al = new ArrayList();
 
+      if (this.threadConnectionMap == null) {
```
[the remark this answers is not in the document]
Comment (mkevo) [reply geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java]: Thanks for comments! Can we just do `synchronized(this.threadConnectionMap)` at the start and not use copy to local variable? In this case we don't need always to check if threadConnectionMap is changed as it can be only changed by this thread.

[the remark this answers is not in the document]
Comment (dschneider-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/tcp/ConnectionTable.java]: I'm not up to speed on this code but I see close() just nulling out this.threadConnectionMap. It does not actually iterator over its contents and close the cons in it. I think this might be happening in another structure: this.threadConnMaps. I would consider changing the code to not null out this.threadConnectionMap. Then you could make this field final and get rid of all the null checks. If we do need to test for it being null then can you use this.closed instead?

**Comment by pivotal-jbarrett** [comment]
> Instead of adding synchronizations and null checks everywhere let's just make that field `final` and change the `close()` method to clear it.

Even butt but I think there is still a race in what to do with calls to this method concurrently with close. The map would be cleared by close but then updated by this method. Having little background in what this is tracking I can’t say if this is good or bad, it is simply just a case the original code seemed to try to avoid.

**Comment by bschuchardt** [comment]
I don't think we should be using the state of a collection as an indication of whether the service is open or closed.  There should be other state that we use to avoid creating a new connection and putting it in the connection map if ConnectionTable is closed.

**Comment by pivotal-jbarrett** [comment]
> I don't think we should be using the state of a collection as an indication of whether the service is open or closed. There should be other state that we use to avoid creating a new connection and putting it in the connection map if ConnectionTable is closed.

Couldn't agree more! If you think its safe for connection references that race into the map on close to sit there then this is an easy fix. A simple check if the collection is closed after the put completes could accomplish this. The close itself would empty the map and any race on that operation would get cleared up on the back end after the put by checking for closed state and removing the entry just put.

**Comment by mhansonp** [comment]
@bschuchardt @pivotal-jbarrett  Are there still concerns with this PR?

**Comment by bschuchardt** [comment]
> @bschuchardt @pivotal-jbarrett Are there still concerns with this PR?

My concern over the threadConnectionMap null checks hasn't been addressed.  I've been fighting against the nulling-out of instance variables like this forever.  It's always causing NPEs for unsuspecting programmers who don't recognize that this anti-pattern is being used.  The instance variable ought to be "final" and some other state should be added and consulted to see if the connection table has been closed.

**Comment by mkevo** [comment]
> > @bschuchardt @pivotal-jbarrett Are there still concerns with this PR?
> 
> My concern over the threadConnectionMap null checks hasn't been addressed. I've been fighting against the nulling-out of instance variables like this forever. It's always causing NPEs for unsuspecting programmers who don't recognize that this anti-pattern is being used. The instance variable ought to be "final" and some other state should be added and consulted to see if the connection table has been closed.

Hi @bschuchardt , 
I think that the best way is to make this threadConnectionMap final and change this `close()` to iterate over the map, close all connections and clear map. In this way, we don't need these null checks.
Writing it to local map and then execute command on it isn't good as `computeIfAbsent()` can throw NPE as we didn't know if someone deleted it as we are checking our local copy.

**Comment by mkevo** [comment]
Tnx @bschuchardt! :)
